src/cargo/ops/cargo_rustc/output_depinfo.rs

```diff
@@ -0,0 +1,74 @@
+use std::collections::HashSet;
+use std::io::{Write, BufWriter};
+use std::fs::File;
+use std::path::{Path, PathBuf};
+
+use ops::{Context, Unit};
+use util::{CargoResult, internal};
+use ops::cargo_rustc::fingerprint;
+
+fn render_filename<P: AsRef<Path>>(path: P, basedir: Option<&str>) -> CargoResult<String> {
+    let path = path.as_ref();
+    let relpath = match basedir {
+        None => path,
+        Some(base) => match path.strip_prefix(base) {
+            Ok(relpath) => relpath,
+            _ => path,
+        }
+    };
+    relpath.to_str().ok_or(internal("path not utf-8")).map(|f| f.replace(" ", "\\ "))
+}
+
+fn add_deps_for_unit<'a, 'b>(deps: &mut HashSet<PathBuf>, context: &mut Context<'a, 'b>,
+    unit: &Unit<'a>, visited: &mut HashSet<Unit<'a>>) -> CargoResult<()>
+{
+    if visited.contains(unit) {
+        return Ok(());
+    }
+    visited.insert(unit.clone());
```

Comment:
You can combine these together as:

```rust
if !visited.insert(unit.clone()) {
    return Ok(())
}
```

I believe `clone` is just copying something like 4 pointers, so it shouldn't cost anything.